drivers/i2c/tca9548a.c

```diff
@@ -0,0 +1,404 @@
+ * Name: tca9548a_write_config
+ *
+ * Description:
+ *   Write to the mux register of TCA9548A.
+ *
+ * Input Parameters:
+ *   priv     - A pointer to the TCA9548A device structure.
+ *   regvalue - The value that will be written.
+ *
+ * Returned Value:
+ *   OK on success or a negative error.
+ *
+ ****************************************************************************/
+
+static int tca9548a_write_config(FAR struct tca9548a_dev_s *priv,
+                                 FAR uint8_t regvalue)
+{
+  struct i2c_config_s iconf;
+  int ret;
+
+  iconf.frequency = CONFIG_TCA9548A_I2C_FREQUENCY;
+  iconf.address   = priv->addr;
+  iconf.addrlen   = 7;
+
+  ret = i2c_write(priv->i2c, &iconf, &regvalue, 1);
+
+  i2cinfo("Write to address 0x%02X; register value: 0x%02x ret: %d\n",
+          priv->addr, regvalue, ret);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: tca9548a_read_config
+ *
+ * Description:
+ *   Read the mux register from TCA9548A.
+ *
+ * Input Parameters:
+ *   priv     - A pointer to the TCA9548A device structure.
+ *   regvalue - A pointer to a buffer into which data will be received.
+ *
+ * Returned Value:
+ *   OK on success or a negative error.
+ *
+ ****************************************************************************/
+
+static int tca9548a_read_config(FAR struct tca9548a_dev_s *priv,
+                                FAR uint8_t *regvalue)
+{
+  struct i2c_config_s iconf;
+  int ret;
+
+  iconf.frequency = CONFIG_TCA9548A_I2C_FREQUENCY;
+  iconf.address   = priv->addr;
+  iconf.addrlen   = 7;
+
+  ret = i2c_read(priv->i2c, &iconf, regvalue, 1);
+
+  i2cinfo("Read from address: 0x%02X; register value: 0x%02x ret: %d\n",
+          priv->addr, *regvalue, ret);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: tca9548a_select_channel
+ *
+ * Description:
+ *   Helper function to select a channel in the TCA9548A. The TCA9548A allows
+ *   to select more than 1 channel at same time, but here we are forcing it
+ *   to only enable a single channel by time. It will avoid collision in case
+ *   where two devices with same address are connected to two channels.
+ *
+ * Input Parameters:
+ *   dev  - Pointer to the (virtual) i2c_master_s.
+ *   val  - The channel to be selected.
+ *
+ * Returned Value:
+ *   OK on success or a negative error.
+ *
+ ****************************************************************************/
+
+static int tca9548a_select_channel(FAR struct tca9548a_dev_s *priv,
+                                   uint8_t val)
+{
+  if (val > TCA9548A_SEL_CH7)
+    {
+      /* channel not existent/supported */
+
+      return -EINVAL;
+    }
+
+  if ((priv->state & (1 << val)) != 0)
+    {
+      /* channel already selected */
+
+      return OK;
+    }
+
+  /* Modify state and write it to the mux. Selecting a channel always enables
+   * the device
+   */
+
+  priv->state = (1 << val);
```

Comment:
It should be overwritten only if tca9548a_write_config(priv, priv->state) is successful.

Reply:
True! Thanks!